Completes the code test for exercises phone-number, word-count, bob, …

jerrysims07 opened the pull request:
…and anagram.

I believe this code is ready for your review. Thank you for giving me a chance to work through it and I look forward to your feedback.

Jerry

Commit:
Completes the code test for exercises phone-number, word-count, bob, and anagram.

stevencwarren commented on ruby/bob.rb:
```diff
@@ -0,0 +1,37 @@
+end
+
+class Remark
+  attr_reader = :remark
```
Is this attr_reader necessary?

stevencwarren commented on ruby/bob.rb:
```diff
@@ -0,0 +1,37 @@
+  end
+
+  def is_question?
+    @remark[-1] == "?"
```
Are there any string methods that could accomplish this that are more readable?

stevencwarren commented on ruby/bob.rb:
```diff
@@ -0,0 +1,37 @@
+  end
+
+  def is_only_numbers?
+    no_words = /[^a-zA-Z]*[?]*/
```
Is there a way with regex to check for only numbers instead of not letters.

jerrysims07 replied:
It is more accurate that I am checking for "has_no_words", so I changed the name of this method. After looking into it, I am not seeing a way to check for "only numbers, punctuation, or whitespace" that is more friendly than this.

stevencwarren commented on ruby/bob.rb:
```diff
@@ -0,0 +1,37 @@
+    @remark = Remark.new(remark)
+    return "Fine. Be that way!" if @remark.is_blank?
+    return "Whoa, chill out!" if @remark.is_shouting?
+    return "Sure." if @remark.is_question?
```
It seems you are including multiple returns, can you format this to only return once?

stevencwarren commented on ruby/anagram.rb:
```diff
@@ -0,0 +1,16 @@
+  def match(candidates)
+    anagrams = []
+    for c in candidates do
+      anagrams.push(c) if (c.downcase != self.root_word.downcase) && (c.downcase.split(//).sort == self.root_word.downcase.split(//).sort)
```
Can you dry this up. There seems to be a lot of duplication. Also, can it be split in multiple lines to be more readable?

stevencwarren commented on ruby/anagram.rb:
```diff
@@ -0,0 +1,16 @@
+
+  def match(candidates)
+    anagrams = []
+    for c in candidates do
```
Is there any specific reason you are using a `for/in` loop? (Hint: look up the difference between `for/in` and `each`)

jerrysims07 replied:
This is a side-effect of me living in PythonLand for awhile. After refreshing myself on the `each`, I chose to use the `Array#select` after all to avoid assigning the empty array to a variable.

stevencwarren commented on ruby/anagram.rb:
```diff
@@ -0,0 +1,16 @@
+  end
+
+  def match(candidates)
+    anagrams = []
```
Can you find away to avoid assigning the empty array to a variable?

stevencwarren commented on ruby/anagram.rb:
```diff
@@ -0,0 +1,16 @@
+    for c in candidates do
+      anagrams.push(c) if (c.downcase != self.root_word.downcase) && (c.downcase.split(//).sort == self.root_word.downcase.split(//).sort)
+    end
+    anagrams
```
Is there a way to utilize ruby's implicit returns to avoid explicity returning the anagrams variable?

stevencwarren commented on ruby/anagram.rb:
```diff
@@ -0,0 +1,16 @@
+class Anagram
+  attr_reader :root_word
```
Is this `attr_reader` necessary? I see that you are creating an instance variable of `@root_word` but calling the `root_word` method in Anagram#match.

stevencwarren commented on javascript/phone-number.js:
```diff
@@ -0,0 +1,42 @@
+    strippedNumber = input.replace(/\W/g, '');
+
+    // Switch statement on the length of strippedNumber.
+    switch(true) {
```
There is some repetition in this switch statement. Can you clean it up a bit?

stevencwarren commented on javascript/word-count.js:
```diff
@@ -0,0 +1,29 @@
+
+    // split the input string on one or more instances of
+    // whitespace, tab, or newline
+    wordsArray = data.trim().split(/[\n\t\s]+/);
```
Some of your regex matchers here are redundant.

stevencwarren commented on javascript/word-count.js:
```diff
@@ -0,0 +1,29 @@
+
+      // Check for string exists as property on object and, if so, send that
+      // value as an array to allow it to be treated as a string literal.
+      if(typeof(resultHash[wordsArray[i]]) == "function"){
```
This `typeof(resultHash[wordsArray[i]]) == "function"` matcher is confusing. Is there a more straightforward way to write the entire if else block to get rid of the nested conditionals?

Commit:
Cleans up the Words.count() function considerably.

- Moves the check for property below the increment work
- Instead of checking for the typeof(), it simply uses isNaN() to determine if expected behavior occurred.
- Simple variable assignment to improve readability.
- Removes redundancy in regex statement (i.e. \s took care of all cases needed)

Commit:
Refactor switch statement to reduce redundancy and improve readability.

Instead of a switch statement comparing to "true", this statement is on the length of the
"StrippedNumber". It turns out there are only three cases to deal with regarding the length: 10,
11, and default.

Commit:
Reformats bob.rb to use native string method, remove unnecessary attr_reader, and return single value.
- Uses ruby's native end_with? method instead rolling our own.
- Builds Bob's reply in a variable before single return statement instead of multiple conditional return statements.
- Removes unnecessary attr_reader on "remark".
- Changes class method from "is_only_numbers" to "has_no_words" to better describe what is actually taking place.

Commit:
Reformat anagram.rb for DRYness, readability, ruby-ness.

- Changed the match method so that a single array is built using the select method instead of the more for...in method that I used before. This also removed the unnecessary step in which I had assigned an empty array to a variable.
- Broke out criteria of a match to their own one-line methods for readability.
- Removed unused attr_reader.

jerrysims07 commented:
@stevencwarren 
Thank you for the feedback. I have pushed up commits to reformat my solutions.
